exonum/src/api/mod.rs

```diff
@@ -295,3 +275,46 @@ pub trait Api {
     /// Used to extend Api.
     fn wire<'b>(&self, router: &'b mut Router);
 }
+
+fn url_fragment<T>(request: &Request, name: &str) -> Result<T, ApiError>
+where
+    T: FromStr,
+    T::Err: ::std::error::Error + Send + Sync + 'static,
+{
+    let params = request.extensions.get::<Router>().unwrap();
+    let fragment = params.find(name).ok_or_else(|| {
+        ApiError::BadRequest(format!("Required parameter '{}' is missing", name))
+    })?;
+    let value: T = T::from_str(fragment).map_err(|e| {
+        ApiError::BadRequest(format!("Invalid '{}' parameter: {}", name, e))
+    })?;
+    Ok(value)
+}
+
+fn optional_param<T>(request: &mut Request, name: &str) -> Result<Option<T>, ApiError>
+where
+    T: FromStr,
+    T::Err: ::std::error::Error + Send + Sync + 'static,
+{
+    let map = request.get_ref::<params::Params>().unwrap();
+    let value = match map.find(&[name]) {
+        Some(&params::Value::String(ref param)) => {
+            let value: T = T::from_str(param).map_err(|e| {
+                ApiError::BadRequest(format!("Invalid '{}' parameter: {}", name, e))
+            })?;
+            Some(value)
+        }
+        _ => None,
+    };
+    Ok(value)
+}
+
+fn required_param<T>(request: &mut Request, name: &str) -> Result<T, ApiError>
```

Comment:
Hmm, I wonder if we can make these functions `pub`. They're quite useful

Reply:
Agree. They could be used within this very PR in testkit integration tests. Also, just an idea (maybe, a stupid one): create a trait encompassing this functions and implement it for `Request`.

Reply:
Replied in https://github.com/exonum/exonum/pull/474#discussion_r167296411

TL;DR: this belongs to an HTTP-API crate, and not to the Exonum per se, so I'd prefer not to expose it, or expose with an "warning, unstable API" in the docs.